File: packages/icoder-sdk/src/resources/runs.ts
/**
 * Phase 6 Gate 4 — Runs / RunHistory / RunTrace resources.
 *
 * This module targets the unified agent-run endpoint
 * (POST /api/v1/agents/{agent_id}/run) shipped in Phase 4-F2, plus the
 * observability endpoints (run_history table — alembic 010; run_trace
 * store — alembic 009). The legacy RuntimeResource in runtime.ts targets
 * the older /api/runtime/* surface and is preserved for backward compat.
 *
 * The unified response envelope (Phase 6 Gate 5) carries trace_url so
 * embedders can deep-link to the frontend RunTrace viewer.
 */

import type { AxiosInstance } from 'axios';
import { requestConfig, type iCoDerRequestOptions } from '../request-options.js';

// ── AgentRunResponse (mirror of backend/app/api/agent_run.py:AgentRunResponse) ──

export interface AgentRunCost {
  amount: number;
  currency: 'CNY' | 'USD' | string;
}

export interface AgentRunRequestInput {
  text: string;
  extra?: Record<string, unknown>;
  documents?: AgentRunSourceDocument[];
  upstream_results?: AgentRunUpstreamResult[];
}

export interface AgentRunSourceDocument {
  document_id: string;
  text: string;
  document_version?: string;
  document_type?: string;
  normalization?: 'none' | 'NFC' | 'NFKC';
}

export interface AgentRunUpstreamResult {
  agent_id: string;
  result: Record<string, unknown>;
  run_id: string;
  schema_ref: string;
  /** Server proof returned with the upstream AgentRunResponse. */
  attestation: string;
}

export interface AgentRunRequest {
  input: AgentRunRequestInput;
  runtime_mode?: 'corti_like_fast' | 'medcoder_deep' | string;
  /** Required when authenticating with client_credentials. */
  purpose_of_use?: 'treatment' | 'payment' | 'healthcare_operations'
    | 'quality_improvement' | 'research' | 'public_health';
  include_trace?: boolean;
  include_evidence?: boolean;
}

export interface AgentRunResponse {
  agent_id: string;
  run_id: string;
  trace_id: string;
  /** Phase 6 Gate 5 — frontend deep-link to RunTrace viewer. */
  trace_url: string;
  runtime_mode: string;
  latency_ms: number;
  cost: AgentRunCost | Record<string, unknown>;
  billing?: Record<string, unknown>;
  summary: string;
  result: Record<string, unknown>;
  schema_ref: string;
  result_attestation: string;
  evidence: Array<Record<string, unknown>>;
  warnings: string[];
  manual_review_required: boolean;
  trace_events: Array<Record<string, unknown>>;
  error: boolean;
  error_reason: string;
}

export interface RunStatusResponse {
  run_id: string;
  status: string;
  terminal: boolean;
  agent_id: string;
  trace_id: string;
  runtime_mode: string;
  latency_ms: number;
  cost_amount: number;
  cost_currency: string;
  error: boolean;
  error_reason?: string | null;
  cancel_reason?: string | null;
  cancelled_at?: string | null;
  cancelled_by_user_id?: string | null;
  created_at?: string | null;
  trace_retention_days: number;
  trace_events_purged_at?: string | null;
  trace_events_purged_count: number;
}

export type RunCancelOutcome =
  | 'ALREADY_COMPLETE'
  | 'CANCELLED'
  | 'RECORDED_ONLY';

export interface RunCancelResponse {
  run_id: string;
  outcome: RunCancelOutcome;
  status: string;
  message: string;
  cancel_reason?: string | null;
  cancelled_at?: string | null;
}

export interface RunTraceTokenRenewResponse {
  run_id: string;
  trace_token: string;
  expires_at: number;
  events_url: string;
  trace_url: string;
  trace_retention_days: number;
}

export interface RunSseEvent {
  event: string;
  data: Record<string, unknown>;
  id?: string;
}

export interface RunStreamRetryOptions {
  /** Total connection attempts, including the first one. */
  maxAttempts?: number;
  initialDelayMs?: number;
  maxDelayMs?: number;
  /** Symmetric delay jitter in the range 0..1. Set 0 for deterministic tests. */
  jitterRatio?: number;
  signal?: AbortSignal;
  lastEventId?: string;
  /** Per-connection timeout, cancellation, headers and additive query fields. */
  requestOptions?: iCoDerRequestOptions;
}

export class RunEventStreamError extends Error {
  constructor(
    message: string,
    readonly httpStatus?: number,
    readonly retryable = false,
    readonly errorCode?: string,
  ) {
    super(message);
    this.name = 'RunEventStreamError';
  }
}

export class RunEventRetentionError extends RunEventStreamError {
  constructor(
    httpStatus: number,
    errorCode: string,
    readonly retentionDays?: number,
  ) {
    super(
      `iCoDer run event cursor is outside retention (HTTP ${httpStatus})`,
      httpStatus,
      false,
      errorCode,
    );
    this.name = 'RunEventRetentionError';
  }
}

async function runEventHttpError(response: Response): Promise<RunEventStreamError> {
  let errorCode: string | undefined;
  let retentionDays: number | undefined;
  try {
    const body = await response.json() as Record<string, unknown>;
    const detail = body.detail;
    if (detail && typeof detail === 'object' && !Array.isArray(detail)) {
      const safeDetail = detail as Record<string, unknown>;
      if (typeof safeDetail.code === 'string' && /^[A-Z0-9_]{1,64}$/.test(safeDetail.code)) {
        errorCode = safeDetail.code;
      }
      if (Number.isInteger(safeDetail.retention_days) && Number(safeDetail.retention_days) > 0) {
        retentionDays = Number(safeDetail.retention_days);
      }
    }
  } catch {
    // Never retain or surface arbitrary response bodies; they may contain PHI.
  }
  if (response.status === 410 && errorCode && (
    errorCode === 'SSE_CURSOR_EXPIRED' ||
    errorCode === 'SSE_TRACE_EXPIRED' ||
    errorCode === 'TRACE_EXPIRED'
  )) {
    return new RunEventRetentionError(response.status, errorCode, retentionDays);
  }
  return new RunEventStreamError(
    `iCoDer run event stream failed (HTTP ${response.status})`,
    response.status,
    false,
    errorCode,
  );
}

async function* parseRunSseStream(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<RunSseEvent> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  const parseFrame = (frame: string): RunSseEvent | undefined => {
    let event = 'message';
    let id: string | undefined;
    const dataLines: string[] = [];
    for (const line of frame.split(/\r?\n/)) {
      if (line.startsWith(':')) continue;
      if (line.startsWith('event:')) event = line.slice(6).trimStart();
      else if (line.startsWith('id:')) id = line.slice(3).trimStart();
      else if (line.startsWith('data:')) dataLines.push(line.slice(5).trimStart());
    }
    if (!dataLines.length) return undefined;
    let data: unknown;
    try {
      data = JSON.parse(dataLines.join('\n'));
    } catch (error) {
      throw new RunEventStreamError('iCoDer returned malformed SSE JSON', undefined, false);
    }
    if (!data || typeof data !== 'object' || Array.isArray(data)) {
      throw new RunEventStreamError('iCoDer returned a non-object SSE event', undefined, false);
    }
    return { event, data: data as Record<string, unknown>, ...(id ? { id } : {}) };
  };

  try {
    for (;;) {
      const { done, value } = await reader.read();
      buffer += decoder.decode(value, { stream: !done });
      buffer = buffer.replace(/\r\n/g, '\n');
      let boundary = buffer.indexOf('\n\n');
      while (boundary >= 0) {
        const frame = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);
        const parsed = parseFrame(frame);
        if (parsed) yield parsed;
        boundary = buffer.indexOf('\n\n');
      }
      if (done) break;
    }
    const trailing = parseFrame(buffer);
    if (trailing) yield trailing;
  } finally {
    reader.releaseLock();
  }
}

function isTerminalRunEvent(event: RunSseEvent): boolean {
  return event.event === 'stream.completed' || event.data.name === 'stream.completed';
}

function validateRetryOptions(options: RunStreamRetryOptions) {
  const maxAttempts = options.maxAttempts ?? 4;
  const initialDelayMs = options.initialDelayMs ?? 250;
  const maxDelayMs = options.maxDelayMs ?? 4000;
  const jitterRatio = options.jitterRatio ?? 0.2;
  if (!Number.isInteger(maxAttempts) || maxAttempts < 1) {
    throw new RangeError('maxAttempts must be a positive integer');
  }
  if (initialDelayMs < 0 || maxDelayMs < 0 || maxDelayMs < initialDelayMs) {
    throw new RangeError('retry delays are invalid');
  }
  if (jitterRatio < 0 || jitterRatio > 1) {
    throw new RangeError('jitterRatio must be between 0 and 1');
  }
  return { maxAttempts, initialDelayMs, maxDelayMs, jitterRatio };
}

function retryDelay(
  retryIndex: number,
  initialDelayMs: number,
  maxDelayMs: number,
  jitterRatio: number,
) {
  const base = Math.min(maxDelayMs, initialDelayMs * (2 ** retryIndex));
  const jitter = 1 + ((Math.random() * 2 - 1) * jitterRatio);
  return Math.max(0, Math.min(maxDelayMs, base * jitter));
}

async function waitForRetry(delayMs: number, signal?: AbortSignal) {
  if (signal?.aborted) throw signal.reason ?? new DOMException('Aborted', 'AbortError');
  if (delayMs === 0) return;
  await new Promise<void>((resolve, reject) => {
    const timer = setTimeout(resolve, delayMs);
    signal?.addEventListener('abort', () => {
      clearTimeout(timer);
      reject(signal.reason ?? new DOMException('Aborted', 'AbortError'));
    }, { once: true });
  });
}

// ── RunHistory (alembic 010) ────────────────────────────────────────────

export interface RunHistoryItem {
  run_id: string;
  agent_id: string;
  agent_ref?: string;
  user_id?: string;
  api_client_id?: string;
  started_at: string;
  finished_at?: string;
  latency_ms: number;
  cost_amount: number;
  cost_currency: string;
  status: 'success' | 'failed' | 'timeout' | string;
  error_reason?: string;
  runtime_mode?: string;
  input_chars?: number;
  output_chars?: number;
}

export interface RunHistoryListResponse {
  items: RunHistoryItem[];
  total: number;
}

// ── RunTrace (alembic 009) ──────────────────────────────────────────────

export interface RunTraceEvent {
  step: string;
  status?: string;
  duration_ms?: number;
  ts?: string;
  metadata?: Record<string, unknown>;
}

export interface RunTraceSummary {
  agent_id: string;
  trace_id: string;
  run_status: string;
  runtime_mode: string;
  latency_ms: number;
  cost: { amount: number; currency: string };
  error: boolean;
  error_reason?: string | null;
  trace_capture_status: string;
  created_at?: string | null;
  review_signal: {
    state: 'required' | 'not_required' | 'not_recorded' | string;
    sources: string[];
    authoritative: false;
  };
}

export interface RunTraceTimelineResponse {
  run_id: string;
  timeline: RunTraceEvent[];
  step_count: number;
  summary: RunTraceSummary;
}

export interface RunTraceRawResponse {
  run_id: string;
  events: RunTraceEvent[];
  summary: RunTraceSummary;
}

// ── A2A v0.3 envelope (mirror of app/icoder/agent_runtime/a2a_facade.py) ──

// ── Resources ───────────────────────────────────────────────────────────

export class RunsResource {
  constructor(private http: AxiosInstance) {}

  /**
   * POST /api/v1/agents/{agent_id}/run — unified agent run facade.
   *
   * Routes to either CodingRuntimeDispatcher (medical-coding-agent) or
   * ProviderRegistry (8 PureLLM agents). Returns the unified envelope
   * with trace_id + trace_url (Phase 6 Gate 5).
   *
   * Idempotency: pass an explicit idempotencyKey to deduplicate. The
   * server-side cache is Phase 7; the client always sends the header.
   */
  run(
    agentId: string,
    body: AgentRunRequest,
    idempotencyKey?: string,
    options?: iCoDerRequestOptions,
  ) {
    const headers: Record<string, string> = {};
    if (idempotencyKey) headers['Idempotency-Key'] = idempotencyKey;
    return this.http.post<AgentRunResponse>(
      `/api/v1/agents/${encodeURIComponent(agentId)}/run`,
      body,
      requestConfig(options, {}, headers),
    );
  }

  /**
   * Convenience: run with text-only input (skips the nested envelope).
   */
  runText(agentId: string, text: string, opts: {
    runtime_mode?: string;
    idempotencyKey?: string;
    purpose_of_use?: AgentRunRequest['purpose_of_use'];
    documents?: AgentRunSourceDocument[];
    upstream_results?: AgentRunUpstreamResult[];
  } = {}, options?: iCoDerRequestOptions) {
    return this.run(agentId, {
      input: {
        text,
        documents: opts.documents,
        upstream_results: opts.upstream_results,
      },
      runtime_mode: opts.runtime_mode,
      purpose_of_use: opts.purpose_of_use,
    }, opts.idempotencyKey, options);
  }

  /** GET /api/v1/runs/{run_id} — authoritative lifecycle state. */
  get(runId: string, options?: iCoDerRequestOptions) {
    return this.http.get<RunStatusResponse>(
      `/api/v1/runs/${encodeURIComponent(runId)}`,
      requestConfig(options),
    );
  }

  /**
   * POST /api/v1/runs/{run_id}/cancel.
   *
   * Callers must inspect ``outcome``. ``RECORDED_ONLY`` means the request was
   * audited but the provider call is still running; it is never reported as a
   * successful cancellation.
   */
  cancel(runId: string, reason = '', options?: iCoDerRequestOptions) {
    return this.http.post<RunCancelResponse>(
      `/api/v1/runs/${encodeURIComponent(runId)}/cancel`,
      { reason },
      requestConfig(options),
    );
  }

  /** Renew a run-bound trace authorization using the configured bearer identity. */
  renewTraceToken(runId: string, options?: iCoDerRequestOptions) {
    return this.http.post<RunTraceTokenRenewResponse>(
      `/api/v1/runs/${encodeURIComponent(runId)}/trace-token`,
      undefined,
      requestConfig(options),
    );
  }

  /**
   * Open the signed, PHI-safe run lifecycle SSE stream.
   *
   * ``traceToken`` is the ``token`` query value from ``AgentRunResponse.trace_url``.
   * It is deliberately supplied separately so this SDK never fetches an arbitrary
   * URL returned by a server.
   */
  async streamEvents(
    runId: string,
    traceToken: string,
    signal?: AbortSignal,
    lastEventId?: string,
    options?: iCoDerRequestOptions,
  ): Promise<ReadableStream<Uint8Array>> {
    if (!traceToken) throw new Error('traceToken is required');
    if (lastEventId && (lastEventId.length > 128 || /[\r\n\0]/.test(lastEventId))) {
      throw new Error('lastEventId is malformed');
    }
    if (options?.maxRetries !== undefined && options.maxRetries !== 0) {
      throw new RangeError('run event streams require maxRetries to be 0');
    }
    if (signal && options?.abortSignal && signal !== options.abortSignal) {
      throw new TypeError('run event stream received conflicting abort signals');
    }
    const baseURL = (this.http.defaults.baseURL || '').replace(/\/$/, '');
    const domainHeaders: Record<string, string> = { Accept: 'text/event-stream' };
    if (lastEventId) domainHeaders['Last-Event-ID'] = lastEventId;
    const config = requestConfig(options, { token: traceToken }, domainHeaders);
    const query = new URLSearchParams();
    for (const [key, value] of Object.entries(config.params ?? {})) {
      if (value !== undefined && value !== null) query.append(key, String(value));
    }
    const signals = [options?.abortSignal ?? signal].filter(
      (value): value is AbortSignal => value !== undefined,
    );
    if (config.timeout !== undefined) signals.push(AbortSignal.timeout(config.timeout));
    const boundedSignal = signals.length > 1 ? AbortSignal.any(signals) : signals[0];
    const response = await fetch(
      `${baseURL}/api/v1/runs/${encodeURIComponent(runId)}/events?${query}`,
      {
        method: 'GET',
        headers: { ...((config.headers ?? {}) as Record<string, string>) },
        signal: boundedSignal,
      },
    );
    if (!response.ok) {
      throw await runEventHttpError(response);
    }
    if (!response.headers.get('content-type')?.startsWith('text/event-stream')) {
      throw new RunEventStreamError('iCoDer returned a non-SSE run event response');
    }
    if (!response.body) throw new RunEventStreamError('iCoDer returned an empty run event stream');
    return response.body;
  }

  /**
   * Yield parsed lifecycle events while reconnecting only bounded transport
   * failures.  The last acknowledged SSE ID is preserved across attempts.
   * A signed-token 401 triggers one bearer-authenticated renewal per failed
   * connection; authorization, cursor and protocol errors are never hidden.
   */
  async *streamEventsResilient(
    runId: string,
    traceToken: string,
    options: RunStreamRetryOptions = {},
  ): AsyncGenerator<RunSseEvent> {
    if (!traceToken) throw new Error('traceToken is required');
    const retry = validateRetryOptions(options);
    let currentToken = traceToken;
    let lastEventId = options.lastEventId;

    for (let attempt = 0; attempt < retry.maxAttempts; attempt += 1) {
      if (options.signal?.aborted) {
        throw options.signal.reason ?? new DOMException('Aborted', 'AbortError');
      }
      try {
        const stream = await this.streamEvents(
          runId,
          currentToken,
          options.signal,
          lastEventId,
          options.requestOptions,
        );
        let completed = false;
        for await (const event of parseRunSseStream(stream)) {
          if (event.id) lastEventId = event.id;
          completed ||= isTerminalRunEvent(event);
          yield event;
        }
        if (completed) return;
        throw new RunEventStreamError(
          'iCoDer run event stream ended before completion', undefined, true,
        );
      } catch (error) {
        if (options.signal?.aborted || (error instanceof Error && error.name === 'AbortError')) {
          throw error;
        }
        const streamError = error instanceof RunEventStreamError ? error : undefined;
        const canRenew = streamError?.httpStatus === 401;
        const retryableTransport = streamError?.retryable || error instanceof TypeError;
        if (!canRenew && !retryableTransport) throw error;
        if (attempt + 1 >= retry.maxAttempts) throw error;
        if (canRenew) {
          const renewed = await this.renewTraceToken(runId, options.requestOptions);
          currentToken = renewed.data.trace_token;
        }
        await waitForRetry(
          retryDelay(attempt, retry.initialDelayMs, retry.maxDelayMs, retry.jitterRatio),
          options.signal,
        );
      }
    }
  }
}

export class RunHistoryResource {
  constructor(private http: AxiosInstance) {}

  /**
   * GET /api/runtime/runs/history — paginated run history.
   *
   * Returns the run_history table (alembic 010). Filter by agent_id,
   * days window, or both.
   */
  list(
    params: { agent_id?: string; days?: number; limit?: number; offset?: number } = {},
    options?: iCoDerRequestOptions,
  ) {
    return this.http.get<RunHistoryListResponse>(
      '/api/runtime/runs/history', requestConfig(options, params),
    );
  }
}

export class RunTraceResource {
  constructor(private http: AxiosInstance) {}

  /**
   * GET /api/runtime/runs/{run_id}/trace?format=timeline
   *
   * Returns the display-safe timeline (sorted by ts). Default format.
   */
  timeline(runId: string, options?: iCoDerRequestOptions) {
    return this.http.get<RunTraceTimelineResponse>(
      `/api/runtime/runs/${encodeURIComponent(runId)}/trace`,
      requestConfig(options, { format: 'timeline' }),
    );
  }

  /**
   * GET /api/runtime/runs/{run_id}/trace?format=raw
   *
   * Returns the raw event dump (for debugging).
   */
  raw(runId: string, options?: iCoDerRequestOptions) {
    return this.http.get<RunTraceRawResponse>(
      `/api/runtime/runs/${encodeURIComponent(runId)}/trace`,
      requestConfig(options, { format: 'raw' }),
    );
  }
}

File: packages/icoder-python/icoder_sdk/resources/runs.py
"""Unified Agent Run and launch-candidate Agent Hub resources."""

from __future__ import annotations

from collections.abc import Mapping
import json
import random
import threading
import time
from typing import Any, Iterator, Optional
from urllib.parse import quote

from ..client import iCoDerClient
from ..request_options import RequestOptions, iCoDerRequestCancelledError
from ..types import (
    AgentHubResponse,
    AgentHubTenantReadinessResponse,
    validate_agent_hub_response,
    validate_agent_hub_tenant_readiness_response,
)


class RunEventStreamError(RuntimeError):
    """Sanitized run-event transport error that never retains the signed URL."""

    def __init__(
        self,
        http_status: Optional[int] = None,
        *,
        retryable: bool = False,
        error_code: Optional[str] = None,
    ) -> None:
        suffix = f" (HTTP {http_status})" if http_status else ""
        super().__init__(f"iCoDer run event stream failed{suffix}")
        self.http_status = http_status
        self.retryable = retryable
        self.error_code = error_code


class RunEventRetentionError(RunEventStreamError):
    """A trace or resume cursor was removed by the server retention policy."""

    def __init__(
        self,
        *,
        error_code: str,
        retention_days: Optional[int] = None,
    ) -> None:
        super().__init__(410, retryable=False, error_code=error_code)
        self.retention_days = retention_days


class RunsResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def run(
        self,
        agent_id: str,
        body: dict[str, Any],
        *,
        idempotency_key: Optional[str] = None,
        request_options: Optional[RequestOptions] = None,
    ) -> dict[str, Any]:
        headers = {"Idempotency-Key": idempotency_key} if idempotency_key else None
        response = self._client.post(
            f"/api/v1/agents/{quote(agent_id, safe='')}/run",
            json=body,
            headers=headers,
            request_options=request_options,
        )
        response.raise_for_status()
        return response.json()

    def run_text(
        self,
        agent_id: str,
        text: str,
        *,
        runtime_mode: Optional[str] = None,
        purpose_of_use: Optional[str] = None,
        idempotency_key: Optional[str] = None,
        include_trace: bool = True,
        include_evidence: bool = True,
        documents: Optional[list[dict[str, Any]]] = None,
        upstream_results: Optional[list[dict[str, Any]]] = None,
        request_options: Optional[RequestOptions] = None,
    ) -> dict[str, Any]:
        return self.run(
            agent_id,
            {
                "input": {
                    "text": text,
                    "extra": {},
                    "documents": list(documents or []),
                    "upstream_results": list(upstream_results or []),
                },
                "runtime_mode": runtime_mode,
                "purpose_of_use": purpose_of_use,
                "include_trace": include_trace,
                "include_evidence": include_evidence,
            },
            idempotency_key=idempotency_key,
            request_options=request_options,
        )

    def get(
        self,
        run_id: str,
        request_options: Optional[RequestOptions] = None,
    ) -> dict[str, Any]:
        """Return the authoritative lifecycle state for a run."""
        response = self._client.get(
            f"/api/v1/runs/{quote(run_id, safe='')}",
            request_options=request_options,
        )
        response.raise_for_status()
        return response.json()

    def cancel(
        self,
        run_id: str,
        reason: str = "",
        request_options: Optional[RequestOptions] = None,
    ) -> dict[str, Any]:
        """Request cancellation; callers must inspect the returned outcome."""
        response = self._client.post(
            f"/api/v1/runs/{quote(run_id, safe='')}/cancel",
            json={"reason": reason},
            request_options=request_options,
        )
        response.raise_for_status()
        return response.json()

    def renew_trace_token(
        self,
        run_id: str,
        request_options: Optional[RequestOptions] = None,
    ) -> dict[str, Any]:
        """Renew a run-bound trace authorization with the bearer identity."""
        response = self._client.post(
            f"/api/v1/runs/{quote(run_id, safe='')}/trace-token",
            request_options=request_options,
        )
        response.raise_for_status()
        payload = response.json()
        if not isinstance(payload, dict) or not payload.get("trace_token"):
            raise RunEventStreamError()
        return payload

    def stream_events(
        self,
        run_id: str,
        trace_token: str,
        *,
        last_event_id: Optional[str] = None,
        request_options: Optional[RequestOptions] = None,
    ) -> Iterator[dict[str, Any]]:
        """Yield signed, PHI-safe run lifecycle event envelopes.

        ``trace_token`` is the ``token`` query value from the Agent Run
        ``trace_url``. Supplying it separately prevents the SDK from fetching
        an arbitrary URL returned by a server.
        """
        if not trace_token:
            raise ValueError("trace_token is required")
        if last_event_id and (
            len(last_event_id) > 128
            or any(char in last_event_id for char in ("\r", "\n", "\x00"))
        ):
            raise ValueError("last_event_id is malformed")
        path = f"/api/v1/runs/{quote(run_id, safe='')}/events"
        headers = {"Accept": "text/event-stream"}
        if last_event_id:
            headers["Last-Event-ID"] = last_event_id
        headers, params, timeout, cancel_event = self._stream_request_options(
            request_options,
            domain_headers=headers,
            domain_params={"token": trace_token},
        )
        self._raise_if_stream_cancelled(cancel_event)
        with self._client.http.stream(
            "GET",
            path,
            params=params,
            headers=headers,
            timeout=timeout,
        ) as response:
            if response.status_code < 200 or response.status_code >= 300:
                error_code = None
                retention_days = None
                try:
                    chunks: list[bytes] = []
                    body_size = 0
                    for chunk in response.iter_bytes():
                        body_size += len(chunk)
                        if body_size > 64 * 1024:
                            chunks = []
                            break
                        chunks.append(chunk)
                    body = json.loads(b"".join(chunks)) if chunks else None
                    detail = body.get("detail") if isinstance(body, dict) else None
                    if isinstance(detail, dict):
                        candidate = detail.get("code")
                        if (
                            isinstance(candidate, str)
                            and 0 < len(candidate) <= 64
                            and all(c.isupper() or c.isdigit() or c == "_" for c in candidate)
                        ):
                            error_code = candidate
                        candidate_days = detail.get("retention_days")
                        if isinstance(candidate_days, int) and candidate_days > 0:
                            retention_days = candidate_days
                except (ValueError, TypeError):
                    # Never retain arbitrary response bodies; they may contain PHI.
                    pass
                if response.status_code == 410 and error_code in {
                    "SSE_CURSOR_EXPIRED", "SSE_TRACE_EXPIRED", "TRACE_EXPIRED",
                }:
                    raise RunEventRetentionError(
                        error_code=error_code,
                        retention_days=retention_days,
                    )
                raise RunEventStreamError(
                    response.status_code,
                    error_code=error_code,
                )
            content_type = response.headers.get("content-type", "")
            if not content_type.startswith("text/event-stream"):
                raise RunEventStreamError(response.status_code)
            data_lines: list[str] = []
            for line in response.iter_lines():
                self._raise_if_stream_cancelled(cancel_event)
                if not line:
                    if data_lines:
                        try:
                            payload = json.loads("\n".join(data_lines))
                        except json.JSONDecodeError as error:
                            raise RunEventStreamError(response.status_code) from error
                        if not isinstance(payload, dict):
                            raise RunEventStreamError(response.status_code)
                        yield payload
                    data_lines = []
                elif line.startswith("data:"):
                    data_lines.append(line[5:].lstrip())
            if data_lines:
                try:
                    payload = json.loads("\n".join(data_lines))
                except json.JSONDecodeError as error:
                    raise RunEventStreamError(response.status_code) from error
                if not isinstance(payload, dict):
                    raise RunEventStreamError(response.status_code)
                yield payload

    def stream_events_resilient(
        self,
        run_id: str,
        trace_token: str,
        *,
        last_event_id: Optional[str] = None,
        max_attempts: int = 4,
        initial_delay: float = 0.25,
        max_delay: float = 4.0,
        jitter_ratio: float = 0.2,
        request_options: Optional[RequestOptions] = None,
    ) -> Iterator[dict[str, Any]]:
        """Yield events with bounded reconnect, cursor resume and token renewal.

        Only transport failures and a stream that closes before the explicit
        ``stream.completed`` marker are retried.  HTTP 401 renews the signed
        trace token through the bearer-authenticated endpoint.  Cursor,
        tenancy, authorization and protocol errors remain visible immediately.
        """
        if not trace_token:
            raise ValueError("trace_token is required")
        if not isinstance(max_attempts, int) or max_attempts < 1:
            raise ValueError("max_attempts must be a positive integer")
        if initial_delay < 0 or max_delay < initial_delay:
            raise ValueError("retry delays are invalid")
        if jitter_ratio < 0 or jitter_ratio > 1:
            raise ValueError("jitter_ratio must be between 0 and 1")

        import httpx

        current_token = trace_token
        cursor = last_event_id
        for attempt in range(max_attempts):
            try:
                completed = False
                for event in self.stream_events(
                    run_id,
                    current_token,
                    last_event_id=cursor,
                    request_options=request_options,
                ):
                    meta = event.get("meta")
                    if isinstance(meta, dict) and meta.get("event_id"):
                        cursor = str(meta["event_id"])
                    completed = completed or event.get("name") == "stream.completed"
                    yield event
                if completed:
                    return
                raise RunEventStreamError(retryable=True)
            except RunEventStreamError as error:
                can_renew = error.http_status == 401
                if not can_renew and not error.retryable:
                    raise
                if attempt + 1 >= max_attempts:
                    raise
                if can_renew:
                    current_token = str(
                        self.renew_trace_token(
                            run_id,
                            request_options=request_options,
                        )["trace_token"]
                    )
            except httpx.TransportError:
                if attempt + 1 >= max_attempts:
                    raise

            delay = min(max_delay, initial_delay * (2 ** attempt))
            delay *= 1 + random.uniform(-jitter_ratio, jitter_ratio)
            if delay > 0:
                bounded_delay = max(0.0, min(max_delay, delay))
                cancel_event = request_options.cancel_event if request_options else None
                if cancel_event is not None:
                    if cancel_event.wait(bounded_delay):
                        raise iCoDerRequestCancelledError("iCoDer request was cancelled")
                else:
                    time.sleep(bounded_delay)

    def _stream_request_options(
        self,
        request_options: Optional[RequestOptions],
        *,
        domain_headers: dict[str, str],
        domain_params: dict[str, Any],
    ) -> tuple[dict[str, str], dict[str, Any], float, Optional[threading.Event]]:
        headers = dict(domain_headers)
        params = dict(domain_params)
        timeout = self._client.config.timeout
        cancel_event: Optional[threading.Event] = None
        if request_options is None:
            return headers, params, timeout, cancel_event
        if not isinstance(request_options, RequestOptions):
            raise TypeError("request_options must be a RequestOptions instance")
        if request_options.max_retries not in (None, 0):
            raise ValueError("run event streams require max_retries to be 0")
        value = request_options.timeout_in_seconds
        if value is not None:
            if (
                not isinstance(value, (int, float))
                or isinstance(value, bool)
                or value <= 0
                or value > 3600
            ):
                raise ValueError(
                    "timeout_in_seconds must be greater than 0 and at most 3600"
                )
            timeout = float(value)
        self._client._merge_request_headers(headers, request_options.headers)
        if not isinstance(request_options.query_params, Mapping):
            raise TypeError("request option query_params must be a mapping")
        for name, query_value in request_options.query_params.items():
            self._client._validate_query_pair(name, query_value)
            if name in params:
                raise ValueError(
                    f"request option query parameter {name} conflicts with a resource parameter"
                )
            params[name] = query_value
        cancel_event = request_options.cancel_event
        if cancel_event is not None and not isinstance(cancel_event, threading.Event):
            raise TypeError("cancel_event must be a threading.Event")
        self._raise_if_stream_cancelled(cancel_event)
        return headers, params, timeout, cancel_event

    @staticmethod
    def _raise_if_stream_cancelled(cancel_event: Optional[threading.Event]) -> None:
        if cancel_event is not None and cancel_event.is_set():
            raise iCoDerRequestCancelledError("iCoDer request was cancelled")


class AgentHubResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def list(
        self,
        use_case: Optional[str] = None,
        request_options: Optional[RequestOptions] = None,
    ) -> AgentHubResponse:
        response = self._client.get(
            "/api/icoder/agents/hub",
            params={"use_case": use_case} if use_case else None,
            request_options=request_options,
        )
        response.raise_for_status()
        return validate_agent_hub_response(response.json())

    def readiness(
        self, request_options: Optional[RequestOptions] = None,
    ) -> AgentHubTenantReadinessResponse:
        response = self._client.get(
            "/api/icoder/agents/hub/readiness",
            request_options=request_options,
        )
        response.raise_for_status()
        return validate_agent_hub_tenant_readiness_response(response.json())

    def get_card(
        self,
        agent_id: str,
        request_options: Optional[RequestOptions] = None,
    ) -> dict[str, Any]:
        response = self._client.get(
            f"/api/icoder/agents/{quote(agent_id, safe='')}/card",
            request_options=request_options,
        )
        response.raise_for_status()
        return response.json()
